File: 03373. Maximize the Number of Target Nodes After Connecting Trees II/solution.go
/**
 * Problem: 3373. Maximize the Number of Target Nodes After Connecting Trees II
 *
 * Difficulty: Hard
 *
 * Language: Go
 *
 * Performance: Runtime - 55 ms (Beats 100%)
 */

func maxTargetNodes(edges1 [][]int, edges2 [][]int) []int {
	// Get the number of nodes in each tree
	nodeCount1, nodeCount2 := len(edges1)+1, len(edges2)+1

	// Initialize adjacency lists for both trees
	graph1, graph2 := make([][]int, nodeCount1), make([][]int, nodeCount2)

	// Pre-allocate capacity for adjacency lists in first tree
	for i := 0; i < nodeCount1; i++ {
		graph1[i] = make([]int, 0, 2)
	}

	// Pre-allocate capacity for adjacency lists in second tree
	for i := 0; i < nodeCount2; i++ {
		graph2[i] = make([]int, 0, 2)
	}

	// Build adjacency list for the first tree
	for _, edge := range edges1 {
		from, to := edge[0], edge[1]

		graph1[from] = append(graph1[from], to)
		graph1[to] = append(graph1[to], from)
	}

	// Build adjacency list for the second tree
	for _, edge := range edges2 {
		from, to := edge[0], edge[1]
		
		graph2[from] = append(graph2[from], to)
		graph2[to] = append(graph2[to], from)
	}

	// Get parity and count of nodes in each parity for first tree
	parity1, count1 := bfs(nodeCount1, graph1)
	// Get parity and count of nodes in each parity for second tree
	_, count2 := bfs(nodeCount2, graph2)

	// Find the best parity count from the second tree
	bestParityCount := count2[0]
	if count2[1] > bestParityCount {
		bestParityCount = count2[1]
	}

	// Calculate result for each possible connection point
	result := make([]int, nodeCount1)
	for i := 0; i < nodeCount1; i++ {
		result[i] = count1[parity1[i]] + bestParityCount
	}

	return result
}

func bfs(nodeCount int, graph [][]int) ([]int, [2]int) {
	// Track parity (even/odd distance) of each node
	parity := make([]int, nodeCount)
	// Count of nodes in each parity group
	countByParity := [2]int{}
	// Track visited nodes
	visited := make([]bool, nodeCount)
	// Queue for BFS traversal
	queue := make([]int, nodeCount)
	head, tail := 0, 0

	// Start BFS from node 0
	queue[tail] = 0
	tail++

	visited[0] = true
	countByParity[0] = 1

	// Process BFS queue
	for head < tail {
		node := queue[head]
		head++
		nodeParity := parity[node]

		// Process all neighbors
		for _, neighbor := range graph[node] {
			if !visited[neighbor] {
				visited[neighbor] = true

				// Neighbor has opposite parity
				neighborParity := nodeParity ^ 1

				parity[neighbor] = neighborParity
				countByParity[neighborParity]++
				
				queue[tail] = neighbor
				tail++
			}
		}
	}

	return parity, countByParity
}

File: 03373. Maximize the Number of Target Nodes After Connecting Trees II/solution.js
/**
 * Problem: 3373. Maximize the Number of Target Nodes After Connecting Trees II
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 378 ms (Beats 100%)
 */

/**
 * Returns max target nodes after connecting two trees at each possible node
 *
 * @param {number[][]} edges1 - First tree edges
 * @param {number[][]} edges2 - Second tree edges
 *
 * @returns {number[]} - Max target nodes count for each connection point
 */
const maxTargetNodes = (edges1, edges2) => {
  // Depth-first search to count nodes by color (even/odd depth)
  const dfs = (node, parent, depth, adjacencyList, colorMap) => {
    // Initialize counter for nodes at even depth (0 if depth is odd, 1 if depth is even)
    let evenNodesCount = 1 - (depth % 2)

    // Mark node's color based on depth (0 for even, 1 for odd)
    colorMap[node] = depth % 2

    // Recursively process all neighbors except parent
    for (const neighbor of adjacencyList[node]) {
      if (neighbor === parent) continue

      evenNodesCount += dfs(neighbor, node, depth + 1, adjacencyList, colorMap)
    }

    return evenNodesCount
  }

  // Build adjacency list and calculate counts of each color
  const buildTree = (edges, colorMap) => {
    // Get total number of nodes in the tree
    const totalNodes = edges.length + 1
    // Create empty adjacency list for each node
    const adjacencyList = Array.from({ length: totalNodes }, () => [])

    // Create undirected graph from edges
    for (const [sourceNode, targetNode] of edges) {
      adjacencyList[sourceNode].push(targetNode)
      adjacencyList[targetNode].push(sourceNode)
    }

    // Count nodes at even depth starting from node 0
    const evenDepthCount = dfs(0, -1, 0, adjacencyList, colorMap)

    // Return counts of nodes at even and odd depths
    return [evenDepthCount, totalNodes - evenDepthCount]
  }

  // Get sizes of both trees
  const tree1Size = edges1.length + 1
  const tree2Size = edges2.length + 1

  // Store color (0/1) for each node in both trees
  const colorMap1 = new Array(tree1Size).fill(0)
  const colorMap2 = new Array(tree2Size).fill(0)

  // Get counts of each color for both trees
  const colorCounts1 = buildTree(edges1, colorMap1)
  const colorCounts2 = buildTree(edges2, colorMap2)

  // Calculate result for each possible connection point
  const result = new Array(tree1Size).fill(0)

  // For each node in tree1, calculate max target nodes
  for (let i = 0; i < tree1Size; i++)
    result[i] =
      colorCounts1[colorMap1[i]] + Math.max(colorCounts2[0], colorCounts2[1])

  return result
}
